src/parse/parse-token-request.ts

```diff
@@ -0,0 +1,299 @@
+    // validation service
+    logger.warn(
+      `Invalid audience in client JWT assertion: Expected: ${EXPECTED_PRIVATE_KEY_JWT_AUDIENCE}, Received: ${parsedClientAssertion.payload.aud}`
+    );
+
+    throw new TokenRequestError({
+      errorCode: "invalid_client",
+      errorDescription: "Invalid signature in private_key_jwt",
+      httpStatusCode: 400,
+    });
+  }
+
+  if (parsedClientAssertion.payload.sub !== parsedClientAssertion.payload.iss) {
+    logger.warn(
+      {
+        iss_claim: parsedClientAssertion.payload.iss,
+        sub_claim: parsedClientAssertion.payload.sub,
+      },
+      "client_assertion iss and sub claims do not match"
+    );
+
+    throw new TokenRequestError({
+      errorCode: "invalid_client",
+      errorDescription: "Invalid signature in private_key_jwt",
+      httpStatusCode: 400,
+    });
+  }
+
+  if (
+    !(await isSignatureValid(
+      clientPublicKey,
+      tokenSigningAlgorithm,
+      parsedClientAssertion.token
+    ))
+  ) {
+    logger.warn("Could not verify signature of private_key_jwt");
+    throw new TokenRequestError({
+      errorCode: "invalid_client",
+      errorDescription: "Invalid signature in private_key_jwt",
+      httpStatusCode: 400,
+    });
+  }
+
+  return {
+    tokenRequest: {
+      grant_type: tokenRequestBody.grant_type,
+      redirectUri: tokenRequestBody.redirect_uri,
+      code: tokenRequestBody.code,
+      client_assertion: tokenRequestBody.client_assertion,
+      client_assertion_type: tokenRequestBody.client_assertion_type,
+    },
+    clientAssertion: parsedClientAssertion,
+  };
+};
+
+const parseClientAssertion = (
+  clientAssertion: string,
+  tokenRequestClientId?: string
+): PrivateKeyJwt => {
+  const jwtParts = clientAssertion.split(".");
+
+  if (jwtParts.length !== 3) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Unexpected number of Base64URL parts, must be three"
+    );
+  }
+
+  const payload = decodeJwt(clientAssertion);
+  const header = decodeProtectedHeader(clientAssertion);
+  const signature = jwtParts[2];
+
+  if (!signature || signature.trim().length === 0) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: The signature must not be empty"
+    );
+  }
+
+  if (!payload.sub) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing subject in client JWT assertion"
+    );
+  }
+
+  if (!payload.iss) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing issuer in client JWT assertion"
+    );
+  }
+
+  if (!payload.aud) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing audience in client JWT assertion"
+    );
+  }
+
+  if (!payload.exp || typeof payload.exp !== "number") {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing or invalid expiry in client JWT assertion"
+    );
+  }
```

Comment:
Do you remember where you found these two? I can't find them

Reply:
There is also some steps in `JWTAssertionDetails.parse` such as where any time based claims (like iat, exp, nbf) are parsed into dates, where if these aren't numbers a parse error is thrown